### PR TITLE
Add tests for keyboard shortcuts

### PR DESCRIPTION
**Pull request by NilayMajorwar**
This PR adds some tests for the keyboard shortcuts logic. In particular, the tests cover the following:
- `getShortcutStringFromEvent`: Parse event object into respective shortcut string
- `isValidShortcutEvent`: Check if an event object is a valid shortcut keypress
- `getElectronAccelerator`: Convert internal shortcut strings to Electron-compatible "accelerator" strings

This is my first time writing tests for an actual project, so it may not be very good :sweat_smile:. I've covered all the exported functions from `src/KeyboardShortcuts/index.js`, except:
- `getShortcutDisplayName`: it's user interface, so isn't necessary to test, right?
- `getShortcutMetadataFromEvent`: this functionality (except reserved shortcuts) is covered in tested functions
- the two React hooks `useShortcutMap` and `useKeyboardShortcuts`

I've also made some very small refactoring changes to `src/KeyboardShortcuts/index.js`, that do not change the logic in any manner.

### CODE REVIEW

File: newIDE/app/src/KeyboardShortcuts/KeyboardShortcuts.spec.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  getShortcutStringFromEvent,
+  isValidShortcutEvent,
+  getElectronAccelerator,
+} from './index';
+
+/**
+ * Creates a KeyboardEvent-like object for testing
+ * functions that take event objects as input
+ */
+const keyEvent = (ctrlKey, shiftKey, altKey, code) => ({
```
Comment (4ian) [inline newIDE/app/src/KeyboardShortcuts/KeyboardShortcuts.spec.js:11]: Uber nitpicking: I usually make functions to always start with a verb, because it's something that "do" something.
Here: `makeKeyEvent`

File: newIDE/app/src/KeyboardShortcuts/KeyboardShortcuts.spec.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
+import {
```
Comment (4ian) [inline newIDE/app/src/KeyboardShortcuts/KeyboardShortcuts.spec.js:1]: Try to add `// @flow` and see if there are any errors.
Even for test, we try to have flow coverage so that we can get static typing and the safety that comes with it when doing refactorings in the future :)

**Comment by 4ian** [comment]
Looks great! Thanks a lot for adding these tests, they are very useful to ensure future refactoring/work on this can be done safely. They are also great "examples" for the reader that tries to understand your code. 

> getShortcutDisplayName: it's user interface, so isn't necessary to test, right?

Depends on the risk associated with it. If it breaks, is it a huge problem? Is it complex enough so that someone could break it in the future and not see it immediately? Are there any edge case that is hard to see and not often encountered, that you would like to show in a test?
If any answer is yes, you can consider making a test :) 

> getShortcutMetadataFromEvent: this functionality (except reserved shortcuts) is covered in tested functions

It's fine if it's mostly covered by others tests. 
As we say "Write tests. Not too many. Mostly integration." - so if your function is tested by calling others, it's fine already.

> the two React hooks useShortcutMap and useKeyboardShortcuts

React hooks would need to be tested inside components - we have almost none of these tests so far. It would be an interesting exercise to do - though I think that these tests using components are often hard to set up, maintain and I prefer making "stories" in Storybook to make a "visual test case" that is useful for the next developer.

Finally, I tend to say that if there is too much logic in a hook or a component, this logic should be moved to external functions, that can be tested (up to a point, as with everything)

So I would say not to worry with all of these. What you did is already great 👍

File: newIDE/app/src/KeyboardShortcuts/KeyboardShortcuts.spec.js (hunk 1 above)
Comment (NilayMajorwar) [reply newIDE/app/src/KeyboardShortcuts/KeyboardShortcuts.spec.js]: My bad - fixed.

File: newIDE/app/src/KeyboardShortcuts/KeyboardShortcuts.spec.js (hunk 2 above)
Comment (NilayMajorwar) [reply newIDE/app/src/KeyboardShortcuts/KeyboardShortcuts.spec.js:1]: I would've done that, but I'll have to force-typecast the output of `makeKeyEvent` (`object` -> `any` -> `KeyboardEvent`) since two of the three tested functions take an event object as argument. The rest of it is fine. Should I do it?

**Comment by NilayMajorwar** [comment]
Yup, that rules out pretty much everything I think. `useShortcutMap` is a nearly trivial hook and `useKeyboardShortcuts`, while looking bulky, is more or less a combination of helper functions already.

On a side note though, if we were to create a Storybook test case for `useKeyboardShortcuts`, it would involve creating a component with:
- a display that shows the command name for the shortcut pressed
- a dummy text input - to test/ensure shortcuts don't work when user is typing
- a dummy dialog - to test/ensure shortcuts don't work when a dialog is open

With some more gears and switches, maybe this test component can act as a really good test case for keyboard shortcuts - checking clashes with user typing and open dialogs, reserved shortcuts, nearly everything. Perhaps I can work on it and add it to this PR?

File: newIDE/app/src/KeyboardShortcuts/KeyboardShortcuts.spec.js (hunk 2 above)
Comment (4ian) [reply newIDE/app/src/KeyboardShortcuts/KeyboardShortcuts.spec.js:1]: Let's still do it and add a `// $FlowIgnore - we intentionally create a fake KeyboardEvent` like comment to it :) 
Better to have a file typed even if there is a comment to ignore an error because it's a test - so not everything is perfectly simulated.

**Comment by 4ian** [comment]
> With some more gears and switches, maybe this test component can act as a really good test case for keyboard shortcuts - checking clashes with user typing and open dialogs, reserved shortcuts, nearly everything. Perhaps I can work on it and add it to this PR?

That could be useful, feel free to take a look, either in in this PR or in another as you prefer 👍
It would be great to have. As storybook is running inside the browser not Electron we can't test everything, but it's already very good if we have a "test bench" to quickly check "is pressing a shortcut working"? "how does it work with inputs"? "how does it work with dialogs?" :)

**Comment by NilayMajorwar** [comment]
I did some work on the test bed, but looks like it's a bit tougher than I thought. So I think I'll take some time and do it in another PR.

**Comment by 4ian** [comment]
Sure, no worries! Looks great, thanks for the PR, very well done 👍
